src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Configuration/ImplicitlyActiveDimensionProvider.cs

```diff
@@ -0,0 +1,64 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using System.Linq;
+
+namespace Microsoft.VisualStudio.ProjectSystem.Configuration
+{
+    /// <summary>
+    ///     Provides an implementation of <see cref="IImplicitlyActiveDimensionProvider"/> that bases
+    ///     itself on <see cref="IProjectConfigurationDimensionsProvider"/> instances.
+    /// </summary>
+    [Export(typeof(IImplicitlyActiveDimensionProvider))]
+    internal class ImplicitlyActiveDimensionProvider : IImplicitlyActiveDimensionProvider
+    {
+        private readonly Lazy<ImmutableArray<string>> _builtInImplicitlyActiveDimensions;
+
+        [ImportingConstructor]
+        public ImplicitlyActiveDimensionProvider(UnconfiguredProject project)
+        {
+            _builtInImplicitlyActiveDimensions = new Lazy<ImmutableArray<string>>(CalculateBuiltInImplicitlyActiveDimensions);
+
+            DimensionProviders = new OrderPrecedenceImportCollection<IProjectConfigurationDimensionsProvider, IConfigurationDimensionDescriptionMetadataView>(projectCapabilityCheckProvider: project);
+        }
+
+        [ImportMany]
+        internal OrderPrecedenceImportCollection<IProjectConfigurationDimensionsProvider, IConfigurationDimensionDescriptionMetadataView> DimensionProviders
+        {
+            get;
+        }
+
+        public IEnumerable<string> GetImplicitlyActiveDimensions(IEnumerable<string> dimensionNames)
+        {
+            Requires.NotNull(dimensionNames, nameof(dimensionNames));
+
+            ImmutableArray<string> builtInDimensions = _builtInImplicitlyActiveDimensions.Value;
+
+            // NOTE: Order matters; this must be in the order in which the providers are
+            // prioritized in 'builtInDimensions', of which Enumerable.Intersect guarantees.
+            return builtInDimensions.Intersect(dimensionNames, StringComparers.ConfigurationDimensionNames);
+        }
+
+        private ImmutableArray<string> CalculateBuiltInImplicitlyActiveDimensions()
+        {
+            var implicitlyActiveDimensions = ImmutableArray.CreateBuilder<string>();
+
+            foreach (Lazy<IProjectConfigurationDimensionsProvider, IConfigurationDimensionDescriptionMetadataView> provider in DimensionProviders)
+            {
+                for (int i = 0; i < provider.Metadata.IsVariantDimension.Length; i++)
+                {
+                    if (provider.Metadata.IsVariantDimension[i])
+                    {
+                        if (!implicitlyActiveDimensions.Contains(provider.Metadata.DimensionName[i], StringComparers.ConfigurationDimensionNames))
```

Comment:
How many of these are there likely to be?  Should we build this with `HashSet` or something instead?

Reply:
Tiny amount 3 currently, maybe 4 if we add a dimension for RID. I've used ImmutableArray also to guarantee order.